src/python/pants/engine/internals/specs_rules.py

```diff
@@ -474,6 +477,53 @@ def __init__(
         )
 
 
+def _handle_ambiguous_result(
+    request: TargetRootsToFieldSetsRequest,
+    result: TargetRootsToFieldSets,
+) -> TargetRootsToFieldSets:
+    assert len(result.targets) > 1
+    field_set_types = [type(field_set) for field_set in result.field_sets]
+
+    # NB: See https://github.com/pantsbuild/pants/pull/15849. We don't want to break clients as
+    # we shift behavior, so we add this temporary hackery here.
+    if (
+        # (We check for 2 targets because 3 would've been ambiguous pre-our-change)
+        len(result.targets) == 2
+        and PexBinaryFieldSet in field_set_types
+        and PythonSourceFieldSet in field_set_types
```

Comment:
I thought for some reason that the option `--run-packaged-firstparty` influenced behavior? If it's set to True, we just always use the `python_source` w/o warning.

Reply:
I'm not following :sweat_smile: `--run-packaged-firstparty` has to do with the behavior of running a `pex_binary` (once we've already spec'd it out).

Reply:
Yeah, you're right. I was coupling the two things incorrectly.

But, when you have a `pex_binary` and `python_source` refer to the same thing, I do think that we want a way to let you safely use the file argument for the `python_source` w/o getting the warning. It's valid to not delete the `pex_binary` target, so that is not adequate as the sole way to fix the deprecation.

I think what we want is a dedicated option that toggles whether an ambiguous file source matches the `pex_binary` vs. `python_source`. It defaults to `pex_binary`, then in 2.15 we switch to default to `python_source`.

Reply:
> But, when you have a pex_binary and python_source refer to the same thing, I do think that we want a way to let you safely use the file argument for the python_source w/o getting the warning.

Well if you want the python_source... oh I see

Reply:
So many new options and deprecations. This is getting tricky

Reply:
Maybe we can fold both behaviors into one option? `--use-old-run-semantics`?